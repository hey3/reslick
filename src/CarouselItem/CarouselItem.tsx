import React from 'react'
import styled from 'styled-components'

type ContainerProps = React.ImgHTMLAttributes<HTMLImageElement> & {
  className?: string
  /** href for component */
  href?: string
}
type PresenterProps = Record<string, unknown>
type Props = ContainerProps & PresenterProps

const DomComponent: React.FC<Props> = ({ className, href, alt, children, ...props }) => {
  return (
    <li className={`${className} carousel-item-root`}>
      <a className="carousel-link-area" href={href} aria-label={props['aria-label']}>
        <figure className="carousel-image-are">
          <picture className="carousel-picture">
            <img className="carousel-image" loading="lazy" alt={alt} {...props} />
          </picture>
        </figure>
      </a>
      <div className="carousel-content">{children}</div>
    </li>
  )
}

const StyledComponent = styled(DomComponent)`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: calc(2.5rem / 2);
  scroll-snap-align: start;

  a {
    align-self: center;

    & > figure {
      position: relative;
      justify-content: center;
      width: inherit;
      height: inherit;
      inline-size: inherit;
      block-size: inherit;
      margin: 0;
      cursor: pointer;
      user-select: none;

      & > picture {
        display: inline-block;
        width: inherit;
        height: inherit;
        inline-size: inherit;
        block-size: inherit;
        border-radius: 1ex;
        box-shadow: 0 1px 2px 0 hsl(0 0% 0% / 20%);
        overflow: hidden;

        & > img {
          display: block;
          width: 100%;
          height: 100%;
          inline-size: 100%;
          block-size: 100%;
          background-image: linear-gradient(to bottom, hsl(210 10% 40%), hsl(210 10% 20%));
          background-color: #525252;
          object-fit: cover;
        }
      }
    }
  }
`

export const CarouselItem: React.FC<ContainerProps> = ({ children, ...props }) => {
  return <StyledComponent {...props}>{children}</StyledComponent>
}
