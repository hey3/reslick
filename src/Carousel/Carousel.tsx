import React from 'react'
import { CarouselItem } from '../CarouselItem'
import styled, { css, keyframes } from 'styled-components'

type ContainerProps = {
  className?: string
  /** Whether to slide automatically  */
  autoSlide?: boolean
  /** Number of slides to display at one time */
  displayCount: number
  /** Image size in slide (horizontal) */
  imageWidth?: string
  /** Image size in slide (vertical) */
  imageHeight?: string
  children: React.ReactElement<typeof CarouselItem> | Array<React.ReactElement<typeof CarouselItem>>
}

type PresenterProps = {
  slideCount: number
  displayCount: number
  imageWidth: string
  imageHeight: string
}

type Props = ContainerProps & PresenterProps

const DomComponent: React.FC<Props> = ({ className, children }) => {
  return (
    <section className={`${className} carousel-root`}>
      <ul className="horizontal-media-scroller" tabIndex={0}>
        {children}
      </ul>
    </section>
  )
}

const StyledComponent = styled(DomComponent)`
  ul {
    margin: 0;
    padding-inline-start: 0;
  }

  .horizontal-media-scroller {
    display: grid;
    grid-auto-flow: column;
    grid-auto-columns: ${props => `calc(100% / ${props.displayCount} - calc(2.5rem / 2))`};
    gap: calc(2.5rem / 2);
    width: ${props => (props.autoSlide ? `calc(100% / ${props.displayCount})` : 'auto')};
    min-width: 100%;
    padding-inline: 2.5rem;
    padding-block: calc(2.5rem / 2);
    overflow-x: ${props => (props.autoSlide ? 'unset' : 'auto')};
    overscroll-behavior-x: contain;
    scroll-snap-type: x proximity;
    outline-offset: -5px;
    animation: ${props =>
      props.autoSlide
        ? css`
            ${autoScroll(props.slideCount, props.displayCount)} ${(props.slideCount -
              props.displayCount) *
            3}s infinite
          `
        : 'unset'};

    &::-webkit-scrollbar {
      display: none;
    }

    &:focus {
      outline: none;
    }

    &:hover {
      animation-play-state: paused;
    }

    & > li > a {
      width: ${props => props.imageWidth};
      height: ${props => props.imageHeight};
      inline-size: ${props => props.imageWidth};
      block-size: ${props => props.imageHeight};
    }
  }
`

const autoScroll = (slideCount: number, displayCount: number) =>
  keyframes`
    ${[...Array(slideCount - displayCount + 1)].map((_, i) => {
      return css`
        ${(100 / (slideCount - displayCount + 1)) * i}% {
          transform: translateX(calc(calc(-100% / ${displayCount} * ${i})));
        }
      `
    })}
  `

export const Carousel: React.FC<ContainerProps> = ({
  children,
  imageWidth = 'auto',
  imageHeight = 'auto',
  ...props
}) => {
  const slideCount = React.Children.count(children)
  const presenterProps: PresenterProps = {
    slideCount,
    displayCount: slideCount - props.displayCount < 0 ? slideCount : props.displayCount,
    imageWidth,
    imageHeight,
  }
  return (
    <StyledComponent {...props} {...presenterProps}>
      {children}
    </StyledComponent>
  )
}
