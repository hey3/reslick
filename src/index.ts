import { Carousel } from './Carousel'
import { CarouselItem } from './CarouselItem'

export default Carousel

export { Carousel, CarouselItem }
