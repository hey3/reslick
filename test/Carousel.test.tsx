import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import { Carousel } from '../src/Carousel'
import { CarouselItem } from '../src/CarouselItem'

describe('Carousel', () => {
  it('renders correctly', () => {
    const defaultWrapper = renderer
      .create(
        <Carousel displayCount={1}>
          <div>test</div>
        </Carousel>
      )
      .toJSON()
    expect(defaultWrapper).toMatchSnapshot()

    const carouselWrapper = renderer.create(
      <Carousel displayCount={1}>
        <CarouselItem />
      </Carousel>
    ).toJSON()
    expect(carouselWrapper).toMatchSnapshot()
  })
})
