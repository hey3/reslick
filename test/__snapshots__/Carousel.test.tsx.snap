// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly 1`] = `
.c0 ul {
  margin: 0;
  padding-inline-start: 0;
}

.c0 .horizontal-media-scroller {
  display: grid;
  grid-auto-flow: column;
  grid-auto-columns: calc(100% / 1 - calc(2.5rem / 2));
  gap: calc(2.5rem / 2);
  width: auto;
  min-width: 100%;
  padding-inline: 2.5rem;
  padding-block: calc(2.5rem / 2);
  overflow-x: auto;
  overscroll-behavior-x: contain;
  -webkit-scroll-snap-type: x proximity;
  -moz-scroll-snap-type: x proximity;
  -ms-scroll-snap-type: x proximity;
  scroll-snap-type: x proximity;
  outline-offset: -5px;
  -webkit-animation: unset;
  animation: unset;
}

.c0 .horizontal-media-scroller::-webkit-scrollbar {
  display: none;
}

.c0 .horizontal-media-scroller:focus {
  outline: none;
}

.c0 .horizontal-media-scroller:hover {
  -webkit-animation-play-state: paused;
  animation-play-state: paused;
}

.c0 .horizontal-media-scroller > li > a {
  width: auto;
  height: auto;
  inline-size: auto;
  -webkit-block-size: auto;
  -ms-flex-block-size: auto;
  block-size: auto;
}

<section
  className="c0 carousel-root"
>
  <ul
    className="horizontal-media-scroller"
    tabIndex={0}
  >
    <div>
      test
    </div>
  </ul>
</section>
`;

exports[`Carousel renders correctly 2`] = `
.c0 ul {
  margin: 0;
  padding-inline-start: 0;
}

.c0 .horizontal-media-scroller {
  display: grid;
  grid-auto-flow: column;
  grid-auto-columns: calc(100% / 1 - calc(2.5rem / 2));
  gap: calc(2.5rem / 2);
  width: auto;
  min-width: 100%;
  padding-inline: 2.5rem;
  padding-block: calc(2.5rem / 2);
  overflow-x: auto;
  overscroll-behavior-x: contain;
  -webkit-scroll-snap-type: x proximity;
  -moz-scroll-snap-type: x proximity;
  -ms-scroll-snap-type: x proximity;
  scroll-snap-type: x proximity;
  outline-offset: -5px;
  -webkit-animation: unset;
  animation: unset;
}

.c0 .horizontal-media-scroller::-webkit-scrollbar {
  display: none;
}

.c0 .horizontal-media-scroller:focus {
  outline: none;
}

.c0 .horizontal-media-scroller:hover {
  -webkit-animation-play-state: paused;
  animation-play-state: paused;
}

.c0 .horizontal-media-scroller > li > a {
  width: auto;
  height: auto;
  inline-size: auto;
  -webkit-block-size: auto;
  -ms-flex-block-size: auto;
  block-size: auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: calc(2.5rem / 2);
  -webkit-scroll-snap-align: start;
  -moz-scroll-snap-align: start;
  -ms-scroll-snap-align: start;
  scroll-snap-align: start;
}

.c1 a {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c1 a > figure {
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: inherit;
  height: inherit;
  inline-size: inherit;
  -webkit-block-size: inherit;
  -ms-flex-block-size: inherit;
  block-size: inherit;
  margin: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 a > figure > picture {
  display: inline-block;
  width: inherit;
  height: inherit;
  inline-size: inherit;
  -webkit-block-size: inherit;
  -ms-flex-block-size: inherit;
  block-size: inherit;
  border-radius: 1ex;
  box-shadow: 0 1px 2px 0 hsl(0 0% 0% / 20%);
  overflow: hidden;
}

.c1 a > figure > picture > img {
  display: block;
  width: 100%;
  height: 100%;
  inline-size: 100%;
  -webkit-block-size: 100%;
  -ms-flex-block-size: 100%;
  block-size: 100%;
  background-image: linear-gradient(to bottom,hsl(210 10% 40%),hsl(210 10% 20%));
  background-color: #525252;
  object-fit: cover;
}

<section
  className="c0 carousel-root"
>
  <ul
    className="horizontal-media-scroller"
    tabIndex={0}
  >
    <li
      className="c1 carousel-item-root"
    >
      <a
        className="carousel-link-area"
      >
        <figure
          className="carousel-image-are"
        >
          <picture
            className="carousel-picture"
          >
            <img
              className="carousel-image"
              loading="lazy"
            />
          </picture>
        </figure>
      </a>
      <div
        className="carousel-content"
      />
    </li>
  </ul>
</section>
`;
